lib/gc/Transforms/MemRefToCPURuntime.cpp

```diff
@@ -96,6 +97,26 @@ struct ConvertMemRefToCPURuntime
               }
             }
           }
+        } else if (isa<memref::AllocOp>(op)) {
+          allocStack.push_back(op);
+        } else if (isa<memref::DeallocOp>(op)) {
+          // fallback alloc / dealloc not in FILO fashion
+          Value deallocMemref = op->getOperands().front();
+          auto aliases = analysis.resolveReverse(deallocMemref);
```

Comment:
Do we really need the aliases for analyzing deallocMemref vs AllocMemref? They'll be exactly same, right?

Reply:
According to `createBufferDeallocationPass`, `deallocMemref` will be the same as `AllocMemref`. But just in case we will have an IR as below, and directly call the `memref-to-cpuruntime pass`. Do you think we can omit this possibility, and do not use alias analysis?
```
func.func @alloc_dealloc_view() {
  // CHECK-LABEL: func @alloc_dealloc_view()
  // CHECK: %[[m0:.*]] = cpuruntime.alloc() : memref<128xi8>
  %c0 = arith.constant 0: index
  %f0 = arith.constant 0.0: f32
  %alloc = memref.alloc() : memref<128xi8>
  %view = memref.view %alloc[%c0][] : memref<128xi8> to memref<32xf32>
  %subview = memref.subview %view[0][16][1] : memref<32xf32> to memref<16xf32>
  // CHECK: cpuruntime.dealloc
  memref.dealloc %subview : memref<16xf32>
  return
}
```

Reply:
Why do we have such case when it is not stack-like?

Reply:
This case is from [Issue#279](https://github.com/intel/graph-compiler/issues/279). After `bufferDeallocation pass`, we will get the following IR, whose `alloc / dealloc` are in `FIFO` fashion. This case should be resolved by [PR#283](https://github.com/intel/graph-compiler/pull/283), with `MergeAlloc pass` enabled, thus the small temp buffers will be merged into a larger buffer. 
It will be better that the stack-like alloc / dealloc to be checked, in case we still have some non-stack like allocations in some complex scenarios.
```
scf.parallel (%arg11) = (%c0_3) to (%c32_4) step (%c16_5) {
      %alloc_18 = memref.alloc() {alignment = 64 : i64} : memref<1x16x4096xbf16>
      %subview = memref.subview %arg2[0, %arg11, 0] [1, 16, 4096] [1, 1, 1] : memref<1x32x4096xbf16> to memref<1x16x4096xbf16, strided<[131072, 4096, 1], offset: ?>>
      %subview_19 = memref.subview %expand_shape[0, %arg11, 0] [1, 16, 4096] [1, 1, 1] : memref<1x32x4096xbf16> to memref<1x16x4096xbf16, strided<[131072, 4096, 1], offset: ?>>
      scf.parallel (%arg12, %arg13, %arg14) = (%c0, %c0, %c0) to (%c1, %c16, %c4096) step (%c1, %c1, %c1) {
        %3 = memref.load %subview[%arg12, %arg13, %arg14] : memref<1x16x4096xbf16, strided<[131072, 4096, 1], offset: ?>>
        %4 = memref.load %subview_19[%arg12, %arg13, %arg14] : memref<1x16x4096xbf16, strided<[131072, 4096, 1], offset: ?>>
        %5 = arith.extf %4 fastmath<contract> : bf16 to f32
        %6 = arith.extf %3 fastmath<contract> : bf16 to f32
        %7 = arith.addf %6, %5 : f32
        %8 = arith.truncf %7 fastmath<contract> : f32 to bf16
        memref.store %8, %alloc_18[%arg12, %arg13, %arg14] : memref<1x16x4096xbf16>
        scf.reduce 
      }
      %alloc_20 = memref.alloc() {alignment = 64 : i64} : memref<1x16x4096xf32>
      scf.parallel (%arg12, %arg13, %arg14) = (%c0, %c0, %c0) to (%c1, %c16, %c4096) step (%c1, %c1, %c1) {
        %3 = memref.load %alloc_18[%arg12, %arg13, %arg14] : memref<1x16x4096xbf16>
        %4 = arith.extf %3 : bf16 to f32
        memref.store %4, %alloc_20[%arg12, %arg13, %arg14] : memref<1x16x4096xf32>
        scf.reduce 
      }
      memref.dealloc %alloc_18 : memref<1x16x4096xbf16>
      %alloc_21 = memref.alloc() {alignment = 64 : i64} : memref<1x16x4096xf32>
      scf.parallel (%arg12, %arg13, %arg14) = (%c0, %c0, %c0) to (%c1, %c16, %c4096) step (%c1, %c1, %c1) {
        %3 = memref.load %alloc_20[%arg12, %arg13, %arg14] : memref<1x16x4096xf32>
        %4 = memref.load %0[%arg12, %arg13, %arg14] : memref<1x16x4096xf32>
        %5 = math.powf %3, %4 : f32
        memref.store %5, %alloc_21[%arg12, %arg13, %arg14] : memref<1x16x4096xf32>
        scf.reduce 
      }
      memref.dealloc %alloc_20 : memref<1x16x4096xf32>
      %alloc_22 = memref.alloc() {alignment = 64 : i64} : memref<1x16xf32>
      scf.parallel (%arg12, %arg13) = (%c0, %c0) to (%c1, %c16) step (%c1, %c1) {
        memref.store %cst, %alloc_22[%arg12, %arg13] : memref<1x16xf32>
        scf.for %arg14 = %c0 to %c4096 step %c1 {
          %3 = memref.load %alloc_21[%arg12, %arg13, %arg14] : memref<1x16x4096xf32>
          %4 = memref.load %alloc_22[%arg12, %arg13] : memref<1x16xf32>
          %5 = arith.addf %3, %4 : f32
          memref.store %5, %alloc_22[%arg12, %arg13] : memref<1x16xf32>
        }
        scf.reduce 
      }
```

Reply:
>  But just in case we will have an IR as below, and directly call the memref-to-cpuruntime pass. Do you think we can omit this possibility, and do not use alias analysis?

I don't think it's valid to dealloc/free the partial buffer by either `view` or `subivew` of `memref.alloc`, please check the statement here: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td#L552-L555

Reply:
Thanks for pointing out this reference. I will remove the alias.

Reply:
> Why do we have such case when it is not stack-like?

Ideally, all the separated memref.alloc within the same scope should be merged into a big chunk by memref-merge pass, this fallback can be a workaround before the memref-merge patch is pulled in or in case there'll be non FILO alloc/dealloc scenarios.

Reply:
The alloc/free used to be FILO (stack like). Is there any recent optimization in upstream changing that?

> Ideally, all the separated memref.alloc within the same scope should be merged into a big chunk by memref-merge pass,

The buffer-merge pass is an optimization and can be turned off at `-O0` in the future. I suppose we should always make this pass safe to run.

Reply:
If we are handling dynamic shaped buffers, the buffer-merge won't work and we still need many small allocations. I would suggest in our pass pipeline that if stack-like allocator is on, we should try to make sure `BufferDeallocationPipeline` maintain a FILO pattern. If stack-like allocator is off, we can enable non-FILO pattern in BufferDeallocationPipeline